### PR TITLE
ENH: Display a dialog to save trans and fids before quitting the coreg app

### LINKED ISSUE #10236: Coreg GUI should suggest to save fiducials and trans if not saved during interactive session
@GuillaumeFavelier when closing the CoregistrationUI and if nothing was saved by the user I would suggest to load a dialog asking the user if he wants to save the fiducials and the trans before closing

cc @hoechenberger

### PR DESCRIPTION
**Pull request by GuillaumeFavelier**
This PR basically rewrites the `closeEvent()` function to allow displaying a dialog in the coreg app in case the user did not save `trans`.

It is still work in progress. In its current state, it works but the code to achieve it needs refactoring. Plus it does not save the fids yet.

Closes #10236

### CODE REVIEW

**Comment by GuillaumeFavelier** [comment]
https://user-images.githubusercontent.com/18143289/152819894-dea04c8d-7b7e-46a2-93a7-998078e67ef7.mp4

File: mne/gui/_coreg.py [hunk 1]
```diff
@@ -226,10 +228,36 @@ def _get_default(var, val):
...
+                        QMessageBox.Save
+                    )
+                if ret == QMessageBox.Save:
+                    self._forward_widget_command(
+                        "save_trans", "set_value", None)
+                else:
+                    assert ret == QMessageBox.Cancel
+            if not self._fids_saved:
+                self._forward_widget_command(
+                    "save_mri_fids", "set_value", None)
```
Comment (GuillaumeFavelier) [inline mne/gui/_coreg.py:257]: I did not find a nice way to integrate that in the GUI API. That is very specific after all

**Comment by GuillaumeFavelier** [comment]
I will extract the changes made to the GUI API in another PR. I also have to work on a notebook alternative and testing.

**Comment by larsoner** [comment]
If you haven't started, I think it's okay to do the notebook separately. The Qt interface is much more commonly used here

**Comment by GuillaumeFavelier** [comment]
Okay then, let's go with Qt first.

**Comment by larsoner** [comment]
@GuillaumeFavelier since we're not releasing until next week sometime, I optimistically marked this for 1.0. Let me know when it's ready to test! From a quick look at the code, it looks reasonable so far

**Comment by larsoner** [comment]
This is the wrong prompt, it should prompt about the fiducials:
```
$ mne coreg -s sample -d subjects --fif MEG/sample/sample_audvis_raw.fif --trans MEG/sample/sample_audvis_raw-trans.fif
```

https://user-images.githubusercontent.com/2365790/154496353-4ebe5f01-bb1f-4642-8366-e2e7c295cf52.mp4


Also from a UI perspective I expected [save/discard/cancel](https://gitlab.gnome.org/GNOME/gimp/-/issues/2188), where save does save-then-exit, discard does exit (without save), and cancel aborts the exit operation (without save) so that the user can do other things.

**Comment by GuillaumeFavelier** [comment]
Thanks for the feedback, I will rework all of that :+1:

**Comment by larsoner** [comment]
During one interaction I got a segfault caused by:
```
  File "/Users/larsoner/python/mne-python/mne/viz/backends/_qt.py", line 64, in func
    callback(button_name)
  File "/Users/larsoner/python/mne-python/mne/gui/_coreg.py", line 1740, in _close_dialog_callback
    assert button_name == "Discard"
AssertionError
```
It was for MRI fiducials, and the button was labeled "Don't save". I think "Discard" is more common in UI designs.

Also, any Qt event handler should be wrapped with @safe_event so that, even if there are bugs like this, the UI does not segfault.

**Comment by larsoner** [comment]
When I load, lock fids, set to uniform scale mode, then set scaling to 99 instead of 100, I get a box saying that the head<->MRI transform is not saved. This is actually not correct -- it is the head<->MRI trans I loaded, what I didn't save was the "scaled subject MRI". So I think you need to track three states:

1. MRI fids modified
2. trans modified
3. MRI scale modified

Another point -- when the dialog comes up, if I click "Save", it correctly pulls up a save dialog. But if I *then* hit "cancel" on the system dialog -- and thus don't save anything -- the UI still exists. I think it should only exit if the user actually saved a file. In this sense, the cancel of the system dialog works like the cancel of the previous dialog, and the only way a user can "lose their work" is if they actually click the "Discard" button.

**Comment by GuillaumeFavelier** [comment]
> the button was labeled "Don't save".

Oh no, on my system the name is `Discard` so this `button.text()` is not standard or the button itself is not.

> So I think you need to track three states

I see

>  the only way a user can "lose their work" is if they actually click the "Discard" button.

Makes sense

**Comment by larsoner** [comment]
> Oh no, on my system the name is Discard so this button.text() is not standard or the button itself is not.

Okay so it's a standard (my version of) macOS thing, that's fine -- just update the `assert` or remove it

**Comment by GuillaumeFavelier** [comment]
Hopefully it's good enough now :pray:

**Review by larsoner — APPROVED** [review]
LGTM +1 for merge when it comes back green

**Comment by larsoner** [comment]
I still got:
```
(base) bunker:MNE-sample-data larsoner$ mne coreg -s sample -d subjects --fif MEG/sample/sample_audvis_raw.fif --trans MEG/sample/sample_audvis_raw-trans.fif 
Traceback (most recent call last):
  File "/Users/larsoner/python/mne-python/mne/viz/backends/_qt.py", line 67, in func
    callback(button_name)
  File "/Users/larsoner/python/mne-python/mne/gui/_coreg.py", line 1765, in _close_dialog_callback
    assert button_name == "Discard"
AssertionError
Fatal Python error: Aborted
```
when I tweaked a fiducial, hit the X to close, then hit "Don't save". I pushed a commit to get rid of the assertion

**Comment by GuillaumeFavelier** [comment]
What is the exact value of `button_name` that is compared to 'Discard' on your system?

**Comment by larsoner** [comment]
I put it in the code commented out

**Comment by larsoner** [comment]
... more specifically:
```
        #     assert button_name in ("Discard", "Don't Save"), button_name
```

**Comment by GuillaumeFavelier** [comment]
> You checked windows to make sure this is correct there?

While testing https://github.com/hoechenberger/mne-installers on Windows, I ran `mne coreg -s sample -d subjects`:

image

So it uses "Discard" too!
---

I also ran into other bugs specific to Windows but now I can debug properly and fix them in another PR.

* crash during `_configure_status_bar`
* hangs while using `_CheckInside`

**Comment by larsoner** [comment]
> I also ran into other bugs specific to Windows but now I can debug properly and fix them in another PR.

Great -- I would say these are high priority since they will probably make the coreg gui unusable on Windows :(

It now works for me on macOS, thanks @GuillaumeFavelier !

**Comment by GuillaumeFavelier** [comment]
I'll work on windows for a while to fix those 👍

**Pull request merged**